File: broodling/invocation.py
"""Caller-facing composition of Broodling's existing lifecycle services."""

from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass
from pathlib import Path

from .abandonment import AbandonmentCoordinator, AttemptRetirement
from .contract import Contract, RequiredEffect
from .disposition import WorkUnitDisposition, WorkUnitDispositionCoordinator
from .entitlement import SourceSubmission
from .errors import SubmissionNotReady
from .identity import WorkReference
from .ingress import ContractIngress, ContractProposalInput
from .provisioning import AttemptProvisioner
from .store import (
    AdmissionDecisionRecord,
    AttemptAbandonmentRecord,
    AttemptRecord,
    BroodlingStore,
    ContractRevisionRecord,
    EntitledSourceRecord,
    WorktreeAssignmentRecord,
    WorkUnitRecord,
)
from .submission import AttemptSubmission, SubmissionCoordinator
from .zeroshot_sdk import ZeroshotSubmitter


@dataclass(frozen=True, slots=True)
class InvocationStatus:
    """Existing durable facts for one exact Contract revision, not a new state."""

    work_unit: WorkUnitRecord
    sources: tuple[EntitledSourceRecord, ...]
    revision: ContractRevisionRecord
    decision: AdmissionDecisionRecord | None
    attempt: AttemptRecord | None
    worktree: WorktreeAssignmentRecord | None
    submission: AttemptSubmission | None
    abandonment: AttemptAbandonmentRecord | None
    disposition: WorkUnitDisposition | None


class Broodling:
    """One explicit work reference through receipt-backed lifecycle disposition.

    The caller owns the store's lifetime and supplies the existing supported
    Zeroshot configuration. This facade adds no execution or retry policy.
    """

    def __init__(
        self,
        store: BroodlingStore,
        submitter: ZeroshotSubmitter,
        workspace_root: Path | str,
    ) -> None:
        self.store = store
        self.submitter = submitter
        self.workspace_root = workspace_root

    def submit(
        self,
        reference: WorkReference,
        propose: Callable[[ContractProposalInput], Contract],
        *,
        repository: Path | str,
        required_effects: tuple[RequiredEffect, ...],
        revision: str = "HEAD",
        additional_sources: tuple[SourceSubmission, ...] = (),
        constructed_by: str = "model_extraction",
    ) -> InvocationStatus:
        """Capture, propose, admit, provision and dispatch one GitHub request.

        Identical source bytes and proposal resolve the same revision/Attempt.
        Use ``resume`` with its revision ID to avoid fetching or proposing again.
        Repository/revision select B1 only when first admitting an Attempt.
        """
        admitted = ContractIngress(self.store).from_github(
            reference,
            propose,
            required_effects=required_effects,
            additional_sources=additional_sources,
            constructed_by=constructed_by,
        )
        return self.resume(
            admitted.revision.contract_revision_id,
            repository=repository,
            revision=revision,
        )

    def resume(
        self,
        contract_revision_id: str,
        *,
        repository: Path | str | None = None,
        revision: str = "HEAD",
    ) -> InvocationStatus:
        """Continue stored authority without re-fetching or changing frozen B1.

        Repository/revision are required only before first Attempt admission.
        Rejection, completion and abandonment return their retained facts;
        this operation never creates a replacement Attempt.
        """
        decision = self.store.admit(contract_revision_id)
        status = self.status(contract_revision_id)
        if not decision.admitted or status.disposition or status.abandonment:
            return status
        attempt = status.attempt
        if attempt is None:
            if repository is None:
                raise SubmissionNotReady(
                    "first Attempt admission requires a repository"
                )
            attempt = AttemptProvisioner(self.store, self.workspace_root).admit(
                contract_revision_id, repository, revision
            )
        submission = SubmissionCoordinator(self.store, self.submitter)
        submitted = submission.record(attempt.attempt_id)
        if submitted is None:
            AttemptProvisioner(self.store, self.workspace_root).provision(
                attempt.attempt_id
            )
            submission.submit(attempt.attempt_id)
        else:
            # Reconcile the frozen request, never reconstruct it from current
            # execution settings. Already-correlated runs need no credentials.
            submission.reconcile(attempt.attempt_id)
        return self.status(contract_revision_id)

    def status(self, contract_revision_id: str) -> InvocationStatus:
        """Inspect pinned lineage without fetching sources or contacting Zeroshot.

        A current Attempt takes precedence over this revision's historical
        Attempts. Otherwise the last admitted Attempt supplies terminal history.
        No later Contract revision is substituted for the requested one.
        """
        # Read one consistent snapshot using the store's existing transaction
        # boundary; no external operation or mutation occurs in this block.
        with self.store._write():
            revision = self.store.get_contract_revision(contract_revision_id)
            attempts = tuple(
                attempt
                for attempt in self.store.list_attempts(revision.work_unit_id)
                if attempt.contract_revision_id == contract_revision_id
            )
            attempt = next(
                (item for item in attempts if item.is_current),
                attempts[-1] if attempts else None,
            )
            attempt_id = None if attempt is None else attempt.attempt_id
            return InvocationStatus(
                work_unit=self.store.get_work_unit(revision.work_unit_id),
                sources=self.store.contract_source_material(contract_revision_id),
                revision=revision,
                decision=self.store.find_admission_decision(contract_revision_id),
                attempt=attempt,
                worktree=(
                    self.store.worktree_assignment(attempt_id) if attempt else None
                ),
                submission=(
                    SubmissionCoordinator(self.store, self.submitter).record(attempt_id)
                    if attempt
                    else None
                ),
                abandonment=self.store.abandonment(attempt_id) if attempt else None,
                disposition=(
                    WorkUnitDispositionCoordinator(self.store, self.submitter).record(
                        attempt_id
                    )
                    if attempt
                    else None
                ),
            )

    def history(self, reference: WorkReference) -> tuple[InvocationStatus, ...]:
        """Discover retained revision handles, including after a lost response.

        Returns oldest revision first, or an empty tuple for unknown references.
        This never acquires sources, runs a proposer, or resolves a new Work Unit.
        """
        return tuple(
            self.status(revision.contract_revision_id)
            for revision in self.store.list_contract_revisions(reference.work_unit_id)
        )

    async def wait(self, attempt_id: str) -> WorkUnitDisposition:
        """Consume the bound result; cancellation detaches without abandonment.

        SUCCEEDED records authorized delivery, not semantic certification.
        Native failure retains abandonment and raises the existing domain error.
        """
        return await WorkUnitDispositionCoordinator(
            self.store, self.submitter
        ).finalize(attempt_id)

    async def stop(self, attempt_id: str, reason: str) -> AttemptRetirement:
        """Abandon and request native stop, preserving existing cleanup refusals.

        Dispatched Attempts raise CessationUnconfirmed and remain quarantined.
        This operation does not delete worktrees or authorize replacement.
        """
        return await AbandonmentCoordinator(self.store, self.submitter).stop(
            attempt_id, reason
        )

File: src/Broodling/Invocation.cs
namespace Broodling;

/// <summary>
/// The explicitly configured execution target for a new Attempt. An existing Attempt is continued
/// only through the target kind its retained resources name; nothing falls back to the other kind.
/// </summary>
public abstract record InvocationTarget
{
    private InvocationTarget() { }

    /// <summary>No-effect work in an owned local worktree, submitted through the pinned SDK bridge and launcher.</summary>
    public sealed record Local(string WorkspaceRoot, NativeProfile Profile, INativeTransport Transport) : InvocationTarget;

    /// <summary>Authorized PR work over HTTP/OECP; no Python, SDK client state, workspace root or launcher.</summary>
    public sealed record Direct(string Origin) : InvocationTarget;
}

/// <summary>One explicit work reference. Ended authority is handed back without automatic replacement.</summary>
public sealed class Invocation(BroodlingStore store, InvocationTarget target)
{
    /// <summary>The store routes on the retained record; an HTTP result needs no bridge transport.</summary>
    public Task<AttemptCompletion> WaitAsync(string attemptId, CancellationToken cancellationToken = default) =>
        store.WaitAsync(attemptId, (target as InvocationTarget.Local)?.Transport, cancellationToken);

    /// <summary>Retained handback needs neither dispatch configuration nor the old workspace.</summary>
    public static bool CanResumeWithoutDispatch(AdmissionStatus status)
    {
        var attempt = status.Attempts.LastOrDefault();
        return status.Decision is { Admitted: false } || attempt is { IsCurrent: false }
            || status.Submissions.Any(submission => submission.AttemptId == attempt?.AttemptId && submission.State == "correlated");
    }

    public async Task<AdmissionStatus> SubmitAsync(WorkReference reference, Func<ContractProposalInput, Contract> propose,
        IEnumerable<RequiredEffect> requiredEffects, string repository, string revision = "HEAD",
        IEnumerable<SourceSubmission>? additionalSources = null, string constructedBy = "model_extraction",
        DispatchCredentials? credentials = null, GitHubIssueSource? source = null, CancellationToken cancellationToken = default)
    {
        var admitted = await store.AdmitGitHubAsync(reference, propose, requiredEffects, additionalSources, constructedBy, source, cancellationToken);
        return await ResumeAsync(admitted.Revision.ContractRevisionId, repository, revision, credentials, cancellationToken);
    }

    public async Task<AdmissionStatus> ResumeAsync(string revisionId, string? repository = null, string revision = "HEAD",
        DispatchCredentials? credentials = null, CancellationToken cancellationToken = default)
    {
        var status = store.Status(revisionId);
        if (status.Decision is null)
        {
            store.Admit(revisionId);
            status = store.Status(revisionId);
        }
        if (CanResumeWithoutDispatch(status)) return status;
        var attempt = status.Attempts.LastOrDefault();
        if (attempt is null)
        {
            if (repository is null) throw new AttemptAdmissionError("A source repository is required before first Attempt allocation.");
            attempt = target switch
            {
                InvocationTarget.Direct => store.AdmitHttpAttempt(revisionId, repository, revision),
                InvocationTarget.Local local => store.AdmitAttempt(revisionId, repository, local.WorkspaceRoot, revision),
                _ => throw new UnsupportedRuntime("The invocation target is unsupported.")
            };
        }
        switch (target)
        {
            case InvocationTarget.Direct direct when attempt.ResourceKind == AttemptRecord.Http:
                // Returns the retained preparation, refusing a different configured origin.
                store.PrepareHttpSubmission(attempt.AttemptId, direct.Origin);
                await store.DispatchHttpAsync(attempt.AttemptId, credentials, cancellationToken);
                break;
            case InvocationTarget.Local local when attempt.ResourceKind == AttemptRecord.Worktree:
                // No reacquisition, B1 selection or materialization once dispatch may have happened.
                if (store.FindSubmission(attempt.AttemptId) is null) store.ProvisionAttempt(attempt.AttemptId);
                await store.DispatchAsync(attempt.AttemptId, local.Profile, local.Transport, cancellationToken);
                break;
            default:
                throw new UnsupportedRuntime("The configured target kind differs from the retained Attempt's resource kind.");
        }
        return store.Status(revisionId);
    }
}
